src/vuln_analysis/utils/source_rpm_downloader.py

```diff
@@ -0,0 +1,731 @@
+            
+        Returns:
+            Path: Full path to cached RPM file
+        """
+        # Create cache filename: package-version-release.src.rpm
+        cache_filename = f"{package['name']}-{package['version']}-{package['rel']}.src.rpm"
+        return self.rpm_cache_dir / cache_filename
+
+    def _find_cached_rpm(self, package, platform_version):
+        """
+        Find cached RPM using the cache index.
+        
+        Args:
+            package: Package dictionary with name, version, rel
+            platform_version: Platform version (e.g., 'el9', 'el10')
+            
+        Returns:
+            Path or None: Path to cached RPM file if found, None otherwise
+        """
+        if self.rpm_cache_dir is None or not self.cache_index:
+            return None
+        
+        package_name = package['name']
+        package_version = package['version']
+        
+        # Exact match with platform version
+        key = (package_name, package_version, platform_version)
+        if key in self.cache_index:                
+            return self.cache_index[key]['filepath']        
+        return None
+
+    def _is_cached(self, package, platform_version):
+        """
+        Check if RPM is already cached.
+        
+        Args:
+            package: Package dictionary with name, version, rel
+            platform_version: Platform version (e.g., 'el9', 'el10')
+            
+        Returns:
+            bool: True if RPM exists in cache
+        """
+        return self._find_cached_rpm(package, platform_version) is not None
+
+    def _download_to_cache(self, package, download_url):
+        """
+        Download RPM and store in cache.
+        
+        Args:
+            package: Package dictionary
+            download_url: URL to download RPM from
+            
+        Returns:
+            bool: True if download and cache successful
+        """
+        if self.rpm_cache_dir is None:
+            return False
+        cache_path = self._get_cache_path(package)
+        try:
+            logger.info(f"Downloading {package['name']} to cache: {cache_path}")
+            data = self.session.get(download_url, timeout=120).content
+            cache_path.write_bytes(data)
+            #logger.info(f"Cached {package['name']} ({len(data)} bytes)")
+            #add file to cache
+            parsed = self._parse_rpm_filename(cache_path.name)
+            if parsed:
+                key = (parsed['name'], parsed['version'], parsed['platform'])
+                self.cache_index[key] = {
+                    'filepath': cache_path,
+                    'release': parsed['release'],
+                    'filename': parsed['filename']
+                }
+            return True
+        except Exception as e:
+            logger.error(f"Failed to download {package['name']} to cache: {e}")
+            return False
+
+    def _copy_from_cache(self, package, work_dir, platform_version):
```

Comment:
@RedTanny  Please use type hints/annotation for an increased readability and maintainability.